Yoast/javascript
# P2 281 add schema validation to the store

## Summary


This PR can be summarized in the following changelog entry:

* Stores the schema block validation results in the yoast-seo/editor redux store.

## Relevant technical choices:

* changed the `valid : boolean` method to `validate: BlockValidationResult[]` in order to return more detailed information than just a fail/pass. We now know for each clientId _why_ the validation has passed / has failed /was skipped.
* changed the validation of innerblocks to correctly identify the valid and invalid cases of required and optional inner blocks.
* moved the `validation` files around, so that the `core` code doesn't depend on stuff in the `instructions` subfolders
* replaced some base classes with interfaces for better abstraction.
* added some 

## Test instructions

This PR can be tested by following these steps:

* Make sure that you test against the `P2-281-add-schema-validation-to-the-store` branch on Premium.
* Create a Post.
* add a Job Posting (not available at this time)
* open the browser console (F12)
* (A)remove a required block (e.g. job title)
* this should trigger validation, and that should cause validation to be stored in the wordpress store.
* the console shows you what is sent to the store, and you match it with what you see on-screen. Storing a validation will log an object to the console, with the following properties:
- name: the block type (e.g. job-title)
- clientId: a random guid, an id generated by the web page.
- issues: an array of issues. e.g. a job posting block requires a job location block; if that's missing, that would be an issue.
- result: a code, indicating the validation result for this block:
```
	Skipped = -2, // this block was skipped during validation, on purpose. if you ever see this value, that's a bug.
	Unknown = -1, // this block doesn't have any code that validates it, so we simply don't know if it's valid or not.
	Valid = 0, // this block (AND all of its innerblocks) are fine.
	Invalid = 1, // this block (OR some of its innerblocks) have a problem.
	MissingAttribute = 2, // this block has required attributes, but these attributes are missing or empty.
	MissingBlock = 3, // this schema block must have a particular inner block, but that block doesn't exist.
	TooMany = 4, // there may be only one of this type of block, but we found more.
```
* you can also view the result in the redux browser; get the Redux Dev Tools addon and navigate to the yoast-seo/editor store, you can see the validation result in the the schema-blocks section. <- this is the main goal of this ticket.
* (B) add the job title back; validation should trigger and it should be happy now (code 0)
* (C) add another job title to the post; this is illegal and may not even be possible without some hacking of the html. you might be able to copy an existing block but be sure to change it's clientId!
The validation object in the console should now read TooMany (4) for the block with name "job-title".

Note that these console.logs will be removed once the "jobs" schema block is ready for release, but for now, we can enjoy the information they provide :-)

## Impact check

* This PR affects the following parts of the plugin, which may require extra testing:
  * this PR is behind the schema-blocks feature toggle, which is off. no effect should be noticable when creating a new post, unless the feature is on.

## UI changes

## Quality assurance


Fixes #
